### DISCUSSION TITLE
Azure Devops, DynamoDB and TestContainers

### DISCUSSION THREAD

**Discussion by dsshep** [testcontainers/testcontainers-dotnet#590] [Q&A] [answered]
Hi,

I have a test containers test, using the `amazon/dynamodb-local` official [AWS DynamoDB image](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.DownloadingAndRunning.html). 

I'm trying to run the test in Azure Devops pipelines, but the container never seems to go healthy and times out after ~4 minutes after the call to `StartAsync`. Subsequently trying to create tables against the container fail.

The container configuration looks something like:

```csharp
var builder = new TestcontainersBuilder<TestcontainersContainer>()
       .WithImage("amazon/dynamodb-local:latest")
       .WithCleanUp(true)
       .WithPortBinding(10001, 8000);

dynamoDbContainer = builder.Build();

await dynamoDbContainer.StartAsync();
```

Is there anything I am missing here? Is there an easy way to attach some logging mechanism to get some insight as to what may be going wrong?

**Comment by HofmeisterAn** [comment, marked as answer]
Does your configuration run on your local machine? You can configure [logging](https://github.com/testcontainers/testcontainers-dotnet/wiki/Custom-Configuration#enable-logging) and forward `stdout` and `stderr` with an output consumer, but somehow `DynamoDBLocal.jar` does not write to `stdout` (at least that is what I noticed in a quick test). Getting the logs otherwise is not yet supported (but I can add it, it is not much to do).

You can disable the resource reaper (`TestcontainersSettings.ResourceReaperEnabled = false`), run your configuration and get the logs with the CLI `docker logs CONTAINER` (set the container name with `WithName()`)?

**Reply by dsshep** [reply]
Thanks for the reply. 

Managed to get some console logging going, but there isn't any great insight:

```
 08:25:52 info: 
       Info Log
 08:25:52 warn: 
       Warning Log
 08:25:52 fail: 
       Error Log
 08:25:52 crit: 
       Critical Log
 08:25:53 info: 
       Searching Docker registry credential in CredHelpers
 08:25:53 info: 
       Searching Docker registry credential in Auths
 08:25:53 info: 
       Searching Docker registry credential in CredsStore
 08:25:53 info: 
       Docker registry credential ghcr.io not found
 08:25:56 info: 
       Docker image ghcr.io/psanetra/ryuk:2021.12.20 created
 08:25:56 info: 
       Docker container 0690aede99eaaba4defb4ba7ffce38e20a398bf3ca313d103319471addfef2d2 created
 08:25:56 info: 
       Start Docker container 0690aede99eaaba4defb4ba7ffce38e20a398bf3ca313d103319471addfef2d2
 08:25:56 info: 
       Searching Docker registry credential in Auths
 08:25:56 info: 
       Searching Docker registry credential in CredHelpers
 08:25:56 info: 
       Searching Docker registry credential in CredsStore
 08:25:56 info: 
       Docker registry credential index.docker.io found
 08:26:08 info: 
       Docker image amazon/dynamodb-local:latest created
 08:26:08 info: 
       Docker container b752cc5f9a4b6191d41568ea09484c4e85b18912c7ed7b2514305098f00e7479 created
 08:26:08 info: 
       Start Docker container b752cc5f9a4b6191d41568ea09484c4e85b18912c7ed7b2514305098f00e7479
 08:29:54 info: 
       Delete Docker container b752cc5f9a4b6191d41568ea09484c4e85b18912c7ed7b2514305098f00e7479
```

The configuration does work locally.

Not too sure how I'd configure the docker logs in the pipeline when running the tests?

**Reply by HofmeisterAn** [reply]
> Managed to get some console logging going, but there isn't any great insight:

That is what I meant by the "`DynamoDBLocal.jar` does not write to `stdout`. Is the log from a pipeline run? Do you use a Microsoft-hosted or private agent?

> Not too sure how I'd configure the docker logs in the pipeline when running the tests?

Set a fix name for the container and run the CLI command after your test (you need to disable the resource reaper, otherwise it will removes the container).

> but the container never seems to go healthy and times out after ~4 minutes after the call to `StartAsync`.

Are you sure `StartAsync` times out? I can't imagine what could cause this at the moment. Very unusual. Especially because it deletes the container, it looks like it is able to dispose it and there is now exception. Is your example configuration above complete?

**Reply by dsshep** [reply]
Thank you for your reply,

>That is what I meant by the "DynamoDBLocal.jar does not write to stdout. Is the log from a pipeline run? Do you use a Microsoft-hosted or private agent?

This is the logging from the pipeline run. The setup works locally.

>Set a fix name for the container and run the CLI command after your test (you need to disable the resource reaper, otherwise it will removes the container).

Is it possible to do this in the pipeline? Would I need to do some kind of powershell step after the test failure?

>Are you sure StartAsync times out? I can't imagine what could cause this at the moment. Very unusual. Especially because it deletes the container, it looks like it is able to dispose it and there is now exception. Is your example configuration above complete?

Not convinced looking at the log. The code to connect to the container is as simple as:

```csharp
var dynamoDbClient = new AmazonDynamoDBClient(new AmazonDynamoDBConfig
{
    RegionEndpoint = RegionEndpoint.EUWest1,
    ServiceURL = $"http://localhost:{port}",
});
```

Then the call that times out is trying to create an initial table:

```csharp
var response = await dynamoDbClient.CreateTableAsync(new CreateTableRequest
{
    TableName = "mytable",
    AttributeDefinitions = new List<AttributeDefinition>
    {
        new AttributeDefinition
        {
            AttributeName = "pk",
            AttributeType = ScalarAttributeType.S,
        }
    },
    KeySchema = new List<KeySchemaElement>
    {
        new KeySchemaElement
        {
            AttributeName = "pk",
            KeyType = KeyType.HASH,
        }
    },
    ProvisionedThroughput = new ProvisionedThroughput
    {
        ReadCapacityUnits = 5,
        WriteCapacityUnits = 5,
    }
});
```

Which just returns a 400:

`Response status code does not indicate success: 400 (Bad Request).`

**Reply by HofmeisterAn** [reply]
Try to assign a random public host port `WithPortBinding(8000, true)` and connect to it (you get the random public host port from the container object with `GetMappedPublicPort(8000)`).

Do not use `localhost`, use the `Hostname` property instead. `localhost` ist not always right. That is probably your issue. Use `$"http://{dynamoDbContainer.Hostname}:{dynamoDbContainer.GetMappedPublicPort(8000)}"`.

**Reply by dsshep** [reply]
I have given that a go on the pipeline, but it's still experiencing the same problem:

```csharp
var builder = new TestcontainersBuilder<TestcontainersContainer>()
    .WithImage("amazon/dynamodb-local:latest")
    .WithCleanUp(true)
    .WithPortBinding(8000, true);

dynamoDbContainer = builder.Build();

await dynamoDbContainer.StartAsync();

var dynamoDbClient = new AmazonDynamoDBClient(new AmazonDynamoDBConfig
{
    RegionEndpoint = RegionEndpoint.EUWest1,
    ServiceURL = $"http://{dynamoDbContainer.Hostname}:{dynamoDbContainer.GetMappedPublicPort(8000)}",
});

// same code to create table
```

Not sure if you have any other ideas? Maybe a repo example in devops may be needed...

**Reply by HofmeisterAn** [reply]
> Maybe a repo example in devops may be needed

Yep, that would be nice. The configuration looks good so far. If you have something I can play around with, I can take a look at it.

**Reply by dsshep** [reply]
I've created a repo [here](https://dev.azure.com/davidsheepy/DynamoDB%20Pipeline/_build/results?buildId=55&view=results). Let me know if you have trouble viewing it.

In the process of putting this together, I noticed that the tests run fine through Visual Studio, but don't seem to be happy in `dotnet test` or `dotnet vstest`... So perhaps not related to the pipeline after all?

**Reply by HofmeisterAn** [reply]
> In the process of putting this together, I noticed that the tests run fine through Visual Studio, but don't seem to be happy in `dotnet test` or `dotnet vstest`... So perhaps not related to the pipeline after all?

Ok, I will take a look at it.

**Reply by HofmeisterAn** [reply]
I am not an AWS (DynamoDB) expert, but I assume your issue relates too: https://stackoverflow.com/a/62061272. I guess you are using  an AWS profile in Visual Studio. The [pipeline](https://dev.azure.com/HofmeisterAn/GitHub/_build/results?buildId=2286&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=b801fef0-4998-5417-6df0-af092da60315) runs fine when I provide the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables: [150fa7c3](https://dev.azure.com/HofmeisterAn/GitHub/_git/testcontainers-dotnet-590/commit/150fa7c31640778c2b6ab63baaef5b8837674196?refName=refs/heads/master&path=/UnitTest1.cs). Dunno if that is the right fix.

**Reply by dsshep** [reply]
That was the problem. Thank you so much, you're a hero!
